### PR TITLE
[ENH, BUG]: added `colliders` and `v_structures` and deprecated `compute_v_structures` in `dag.py`

### PR DESCRIPTION
**Pull request by Schefflera-Arboricola**
closes https://github.com/networkx/networkx/issues/7385

<strike>
previously `compute_v_structures` was computing colliders. Now, I've renamed `compute_v_structures` to `compute_colliders`(and updated its implementation too) and added the correct implementation for `compute_v_structures`. Also, updated and added the relevant docs and tests.
</strike>

Thank you :)

### CODE REVIEW

**Comment by Schefflera-Arboricola** [comment]
Need to sort parents because doctests were failing.

<details>
  <summary>Doctest logs</summary>

```
=================================== FAILURES ===================================
_____________ [doctest] networkx.algorithms.dag.compute_colliders ______________
1242     Raises
1243     ------
1244     NetworkXNotImplemented
1245         If `G` is an undirected graph.
1246 
1247     Examples
1248     --------
1249     >>> G = nx.DiGraph()
1250     >>> G.add_edges_from([(1, 2), (0, 5), (3, 1), (2, 4), (3, 1), (4, 5), (1, 5)])
1251     >>> sorted(nx.compute_colliders(G))
Expected:
    [(0, 5, 1), (0, 5, 4), (4, 5, 1)]
Got:
    [(0, 5, 4), (1, 5, 0), (1, 5, 4)]

/home/runner/work/networkx/networkx/networkx/algorithms/dag.py:1251: DocTestFailure
____________ [doctest] networkx.algorithms.dag.compute_v_structures ____________
1284     Raises
1285     ------
1286     NetworkXNotImplemented
1287         If `G` is an undirected graph.
1288 
1289     Examples
1290     --------
1291     >>> G = nx.DiGraph()
1292     >>> G.add_edges_from([(1, 2), (0, 5), (3, 1), (2, 4), (3, 1), (4, 5), (1, 5), (5, 2)])
1293     >>> sorted(nx.compute_v_structures(G))
Expected:
    [(0, 5, 1), (0, 5, 4), (4, 5, 1)]
Got:
    [(0, 5, 4), (1, 5, 0), (1, 5, 4)]

/home/runner/work/networkx/networkx/networkx/algorithms/dag.py:1293: DocTestFailure
```
</details>

File: networkx/algorithms/dag.py [hunk 1]
```diff
@@ -1223,37 +1224,86 @@ def dag_to_branching(G):
...
+    `Pearl's PRIMER, Ch-2 page 50: v-structures def. <http://bayes.cs.ucla.edu/PRIMER/primer-ch2.pdf>`_
     """
-    for collider, preds in G.pred.items():
-        for common_parents in combinations(preds, r=2):
-            # ensure that the colliders are the same
-            common_parents = sorted(common_parents)
-            yield (common_parents[0], collider, common_parents[1])
+    for collider in compute_colliders(G):
+        if not (
+            G.has_edge(collider[0], collider[2]) or G.has_edge(collider[2], collider[0])
```
Comment (dschult) [inline networkx/algorithms/dag.py:1307]: maybe split the collider variable to ease the logic.
```
    for u, c, v in compute_colliders(G):
        if not (u in G[v] or v in G[u]):
```

File: networkx/algorithms/dag.py [hunk 2]
```diff
@@ -1223,37 +1224,86 @@ def dag_to_branching(G):
...
+    --------
+    compute_v_structures
+
+    Notes
+    -----
+    `Wikipedia: Collider in causal graphs <https://en.wikipedia.org/wiki/Collider_(statistics)>`_
+    """
+    for node in G.nodes:
+        for p1, p2 in combinations(G.predecessors(node), 2):
+            p1, p2 = sorted([p1, p2])  # sorting parents for consistent output
```
Comment (dschult) [inline networkx/algorithms/dag.py:1264]: We can't sort nodes.  They can be unsortable when two nodes can't be compared:  
`G=nx.Graph([(1, 'a')])`   or a more cool case where you use graphs as nodes in another graph.

Luckily `combinations` produces p1, p2 in a order (the order of `G.predecessors(node)`). So I think you can remove the line with the sort.  If you take the `sorted` out of the tests too, you can do something like this to avoid the sorting.
```python
    >>> G = nx.DiGraph()
    >>> G.add_edges_from([(1, 2), (0, 5), (3, 1), (2, 4), (3, 1), (4, 5), (1, 5)])
    >>> colliders = nx.compute_colliders(G)
    >>> all(u in G.pred[c] and v in G.pred[c] for u, c, v in colliders)
    True
```

File: networkx/algorithms/dag.py [hunk 3]
```diff
@@ -1223,37 +1224,86 @@ def dag_to_branching(G):
 
 @not_implemented_for("undirected")
 @nx._dispatchable
-def compute_v_structures(G):
-    """Iterate through the graph to compute all v-structures.
+def compute_colliders(G):
+    """In a DAG, if you have three nodes A, B, and C, and there are edges from A to C
+    and from B to C, then C is a collider. This means that both A and B are "causing" C
+    in the graph, and conditioning on C would open up an association between A and B.
```
Comment (dschult) [inline networkx/algorithms/dag.py:1230]: The first line of a doc_string is used to describe the hyperlink in the docs for that function. So it should be a short (one-line) description followed by a blank line and then include this paragraph.  

Something like:
```suggestion
    """Yields 3-tuples of nodes that represent colliders in G
    
    In a DAG, if you have three nodes A, B, and C, and there are edges from A to C
    and from B to C, then C is a collider. This means that both A and B are "causing" C
    in the graph, and conditioning on C would open up an association between A and B.
```

File: networkx/algorithms/dag.py [hunk 4]
```diff
@@ -1223,37 +1224,86 @@ def dag_to_branching(G):
...
+
+    Parameters
+    ----------
+    G : graph
+        A networkx `DiGraph`.
 
-    V-structures are triples in the directed graph where
-    two parent nodes point to the same child and the two parent nodes
-    are not adjacent.
+    Returns
```
Comment (dschult) [inline networkx/algorithms/dag.py:1237]: I guess we are using the `Yields` section heading instead of `Returns` now for generators docstrings.

File: networkx/algorithms/dag.py [hunk 5]
```diff
@@ -1223,37 +1224,86 @@ def dag_to_branching(G):
...
+    for node in G.nodes:
+        for p1, p2 in combinations(G.predecessors(node), 2):
+            p1, p2 = sorted([p1, p2])  # sorting parents for consistent output
+            yield ((p1, node, p2))
+
+
+@not_implemented_for("undirected")
+@nx._dispatchable
+def compute_v_structures(G):
+    """Colliders are triples in the directed graph where two parent nodes point to the
```
Comment (dschult) [inline networkx/algorithms/dag.py:1271]: Similar intro first line in this docstring.

File: networkx/algorithms/dag.py [hunk 6]
```diff
@@ -1223,37 +1224,86 @@ def dag_to_branching(G):
...
+    Raises
+    ------
+    NetworkXNotImplemented
+        If `G` is an undirected graph.
 
     Examples
     --------
     >>> G = nx.DiGraph()
-    >>> G.add_edges_from([(1, 2), (0, 5), (3, 1), (2, 4), (3, 1), (4, 5), (1, 5)])
+    >>> G.add_edges_from([(1, 2), (0, 5), (3, 1), (2, 4), (3, 1), (4, 5), (1, 5), (5, 2)])
```
Comment (dschult) [inline networkx/algorithms/dag.py:1293]: Why are there multiedges in this example? And why did you add an edge to what was there? It seems longer than needed. But I'm probably missing something.

**Review by dschult — COMMENTED** [review]
If we're going to change this much, maybe we should consider changing the name so it doesn't include `compute`. We don't usually include words like that in the function names. What do you think about `v_structures(G)` and `colliders(G)`?  

I'd love to have a better name that `v_structures`   :)

File: networkx/algorithms/dag.py (hunk 6 above)
Comment (Schefflera-Arboricola) [reply networkx/algorithms/dag.py:1297]: Adding the edge (5, 2) creates a collider that is not a v-structure; otherwise, both functions return the same output. I think this just makes the distinction more clear.

**Review by dschult — COMMENTED** [review]
I took the liberty of pushing up some changes that remove both sorting and mins. It works locally so maybe will work on CI.

Removing both the sorting and the min of parents seems to work. The triplets are formed in a deterministic way -- so its the same each time it is run. So I guess this unsorted way of testing doesn't allow other correct solutions, but it does raise if the code no longer works the same way -- and it doesn't encourage users to include lots of sorting and min function calls in their own calls.

In doing this, I noticed that the examples are no longer DAGs. (which led to more investigation)
The docs say these functions should have DAGs as inputs. But it doesn't seem to be checked anywhere. That means we are returning colliders even for directed digraphs with cycles. For better or for worse, there are papers on possibly cyclic causal directed graphs. 
https://www.cmu.edu/dietrich/philosophy/docs/tech-reports/68_Richardson.pdf
https://www.sciencedirect.com/science/article/pii/S0888613X19301410
https://arxiv.org/pdf/1309.6836.pdf
And they do talk about colliders in these situations.

So I think we should leave it possible to use these functions on cyclic graphs. But we should add a warning in the doc_string to check that the graph is acyclic if that is what they want.

I have changed the examples to be DAGs and added a check to the example codes to show that it is a good idea to check for a DAG if you want to restrict to that formulation of a collider. And I've added references and notes to the doc_string.

Please change any of this you want

**Comment by Schefflera-Arboricola** [comment]
Thanks for adding the points about directed cyclic graphs. 

And about the sort and min-max on parents, I think we should keep them so that the tests pass for all OS, because right now the [dispatch test](https://github.com/networkx/networkx/actions/runs/8747951431/job/24007201339?pr=7398) is failing. Also ref. [this comment](https://github.com/networkx/networkx/pull/7398#discussion_r1567337466), here also the doctests in dispatch were failing and (ubuntu, 3.11)'s doctests were also failing in CI(from past commits in this PR) bcoz we didn't use sorted and min-max. LMK what u think. Thx!

File: networkx/algorithms/dag.py [hunk 7]
```diff
@@ -1290,20 +1294,19 @@ def compute_v_structures(G):
     Examples
     --------
     >>> G = nx.DiGraph()
-    >>> G.add_edges_from([(1, 2), (0, 5), (3, 1), (2, 4), (3, 1), (4, 5), (1, 5), (5, 2)])
-    >>> sorted(nx.compute_v_structures(G))
-    [(0, 5, 1), (0, 5, 4), (1, 5, 4)]
+    >>> G.add_edges_from([(1, 2), (0, 4), (2, 3), (3, 4), (1, 4), (4, 2)])
+    >>> vstrct = [(min(p1, p2), v, max(p1, p2)) for p1, v, p2 in v_structures(G)]
+    >>> sorted(vstrct)
```
[the remark this answers is not in the document]
Comment (dschult) [reply networkx/algorithms/dag.py]: Yes -- I see these errors and they are not errors on the other tests. I thought the order of nodes would be determined by the order of the edges provided to nx.DiGraph.  What is different in this Dispatch test environment that reverses the node order? hmmm...   I'd like to investigate further. I'm more bothered by not understanding what is going on than by including sorting. So if I don't find anything we should just put the sorts back in.

[the remark this answers is not in the document]
Comment (Schefflera-Arboricola) [reply networkx/algorithms/dag.py]: it's not just the dispatching test. The (ubuntu,3.11) test also failed(ref. 1st wf failure from the above comment). The dispatching test also uses the same config i.e. ubuntu-latest and python 3.11. So, it might not have anything to do with dispatching.

> yes using `sorted` is necessary, ref the doctest failures(from the recent github wf failure) below(ref. https://github.com/networkx/networkx/actions/runs/8687244560/job/23820377679):
> 
> ```
> =================================== FAILURES ===================================
> _________________ [doctest] networkx.algorithms.dag.colliders __________________
> 1245     ------
> 1246     NetworkXNotImplemented
> 1247         If `G` is an undirected graph.
> 1248 
> 1249     Examples
> 1250     --------
> 1251     >>> G = nx.DiGraph()
> 1252     >>> G.add_edges_from([(1, 2), (0, 4), (2, 3), (3, 4), (1, 4), (4, 2)])
> 1253     >>> colliders = [(min(p1, p2), v, max(p1, p2)) for p1, v, p2 in nx.colliders(G)]
> 1254     >>> colliders
> Expected:
>     [(0, 4, 1), (0, 4, 3), (1, 2, 4), (1, 4, 3)]
> Got:
>     [(1, 2, 4), (0, 4, 3), (0, 4, 1), (1, 4, 3)]
> 
> /home/runner/work/networkx/networkx/networkx/algorithms/dag.py:1254: DocTestFailure
> ________________ [doctest] networkx.algorithms.dag.v_structures ________________
> 1290     ------
> 1291     NetworkXNotImplemented
> 1292         If `G` is an undirected graph.
> 1293 
> 1294     Examples
> 1295     --------
> 1296     >>> G = nx.DiGraph()
> 1297     >>> G.add_edges_from([(1, 2), (0, 4), (2, 3), (3, 4), (1, 4), (4, 2)])
> 1298     >>> vstructures = [(min(p1, p2), v, max(p1, p2)) for p1, v, p2 in nx.v_structures(G)]
> 1299     >>> vstructures
> Expected:
>     [(0, 4, 1), (0, 4, 3), (1, 4, 3)]
> Got:
>     [(0, 4, 3), (0, 4, 1), (1, 4, 3)]
> 
> /home/runner/work/networkx/networkx/networkx/algorithms/dag.py:1299: DocTestFailure
> ```

**Review by rossbar — COMMENTED** [review]
Note that removing `compute_v_structures` in this way will be a breaking change for users (e.g. the user who reported #7385).

IMO the easiest way to deal with this would be to reintroduce a `compute_v_structures` function that essentially just raises a deprecation warning then calls the new `colliders`.

It may also be worth discussing whether these functions should indeed be exported to the main namespace. From a readthrough, I understand that they are intended to work on any acyclic graphs (not only DAGs); however, I don't know how grokkable they are in a general context. IMO `nx.dag.v_structures` is a lot clearer than `nx.v_structures`...

**Comment by Schefflera-Arboricola** [comment]
> IMO the easiest way to deal with this would be to reintroduce a `compute_v_structures` function that essentially just raises a deprecation warning then calls the new `colliders`.

+1 on this.

> It may also be worth discussing whether these functions should indeed be exported to the main namespace. From a readthrough, I understand that they are intended to work on any acyclic graphs (not only DAGs); however, I don't know how grokkable they are in a general context. IMO `nx.dag.v_structures` is a lot clearer than `nx.v_structures`...

I'm not sure by "these functions" if you meant all the functions in `dag.py` or just these 2 functions. But, I also think having `nx.dag.` is better than `nx.` for most of these functions. And I was going through the functions in dag.py and 

- `descendants`, `ancestors`, `all_topological_sorts`, `is_aperiodic`, `transitive_closure`,  seemed like general functions(not specific to DAGs)
- `is_directed_acyclic_graph` can have a `not_implemented_for("undirected")` and `not_implemented_for("multigraph")` decorators instead of checking `G.is_directed()` inside the function
- rest all functions seemed dag-specific

**Comment by dschult** [comment]
With #7432 merged, this should be rebased-on/merged-with main and will hopefully pass the dispatch tests now that nx-loopback preserves edge order when converting during the tests. I don't think there will be any conflicts on the merge, but if it is messy let me know and I'll take a stab at it. (`git merge --abort` is helpful if it gets messy)

**Comment by dschult** [comment]
What I think this discussion is leading toward is the following:
- the `compute_v_structures` function should appear in the main namespace raise a deprecation warning and call `nx.dag.colliders`
- the two new function names `colliders` and `v_structures` should not appear in the main namespace, and instead appear only as `nx.dag.colliders()` and `nx.dag.v_structures()`. That is, `colliders` and `v_structures` should not appear in the `__all__` dunder variable of `dag.py` .
- all three functions should show up in the docs -- and the old `compute_v_structures` doc_string should include a deprecation statement in its doc_string.

**Comment by Schefflera-Arboricola** [comment]
The newly added algos `v_structures` and `colliders` is in the `nx.dag` namespace.

```.py
>>> import networkx as nx
>>> G = nx.DiGraph([(1, 2), (0, 4), (3, 1), (2, 4), (0, 5), (4, 5), (1, 5)])
>>> nx.v_structures(G)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'networkx' has no attribute 'v_structures'
>>> nx.dag.v_structures(G)
<generator object v_structures at 0x1079c9d50>
>>> nx.compute_v_structures(G)
<stdin>:1: DeprecationWarning: 

`compute_v_structures` actually yields colliders. It will be
removed in version 3.6. Use `nx.dag.v_structures` instead.

<generator object colliders at 0x10799b140>
>>> nx.dag.colliders(G)
<generator object colliders at 0x10799b740>
>>> nx.colliders(G)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'networkx' has no attribute 'colliders'
```

Here, I can see the warning locally on my laptop with `stacklevel=5` and not `stacklevel=2`. But that was also the case when I was deprecating the `core.py` algorithms in PR https://github.com/networkx/networkx/pull/7121 and I used `stacklevel=5` there. But, recently, I tried running the following in colab with networkx3.3 and I wasn't able to see the warning:

```.py
import networkx as nx

G = nx.Graph([(1,2),(3,4)])
cn = nx.core_number(G)
G = nx.MultiGraph(G)
print(nx.k_core(G, core_number=cn))

# Output: MultiGraph with 4 nodes and 2 edges
```

but I'm able to see it locally: 

```.py
>>> print(nx.k_core(G, core_number=cn))
<stdin>:1: DeprecationWarning: 

`k_core` will not accept `MultiGraph` objects in version 3.5.
Convert it to an undirected graph instead, using::

	G = nx.Graph(G)

MultiGraph with 4 nodes and 2 edges
```

File: networkx/algorithms/dag.py [hunk 8]
```diff
@@ -1224,36 +1224,189 @@ def dag_to_branching(G):
...
+           a general SAT-based procedure", UAI'13: Proceedings of the Twenty-Ninth
+           Conference on Uncertainty in Artificial Intelligence, pg 301–310,
+           `doi:10.5555/3023638.3023669 <https://dl.acm.org/doi/10.5555/3023638.3023669>`_
+    """
+    import warnings
+
+    warnings.warn(
+        (
+            "\n\n`compute_v_structures` actually yields colliders. It will be\n"
+            "removed in version 3.6. Use `nx.dag.v_structures` instead.\n"
```
Comment (dschult) [inline networkx/algorithms/dag.py:1288]: similarly we could/should add a pointer to `nx.dag.colliders` here too. we don't know which the user actually wanted.

File: networkx/algorithms/tests/test_dag.py [hunk 9]
```diff
@@ -775,3 +775,51 @@ def test_compute_v_structures():
...
+
+def test_compute_v_structures_deprecated():
+    G = nx.DiGraph()
+    with pytest.deprecated_call():
+        nx.compute_v_structures(G)
+
+
+def test_v_structures_raise():
+    G = nx.Graph()
+    pytest.raises(nx.NetworkXNotImplemented, nx.dag.v_structures, G)
```
Comment (dschult) [inline networkx/algorithms/tests/test_dag.py:788]: context manager for testing raises is preferred now. It shows the code in a more readable way.
```suggestion
    with pytest.raises(nx.NetworkXNotImplemented, match="for undirected type"):
        nx.dag.v_structures(G)
```

File: networkx/algorithms/tests/test_dag.py [hunk 10]
```diff
@@ -775,3 +775,51 @@ def test_compute_v_structures():
...
+    assert v_structs == {("A", "B", "C")}
+
+    edges = [(0, 1), (2, 1), (0, 2)]  # adjacent parents case: issues#7385
+    G = nx.DiGraph(edges)
+    assert set(nx.dag.v_structures(G)) == set()
+
+
+def test_colliders_raise():
+    G = nx.Graph()
+    pytest.raises(nx.NetworkXNotImplemented, nx.dag.colliders, G)
```
Comment (dschult) [inline networkx/algorithms/tests/test_dag.py:811]: context manager raises here too?

**Review by dschult — APPROVED** [review]
This looks good to me. I made comments below about using the `pytest.raises` context manager approach to testing exceptions. But those are minor changes.

I approve this PR.  Thanks @Schefflera-Arboricola !

File: networkx/algorithms/dag.py [hunk 11]
```diff
@@ -1224,36 +1224,190 @@ def dag_to_branching(G):
...
+    import warnings
+
+    warnings.warn(
+        (
+            "\n\n`compute_v_structures` actually yields colliders. It will be\n"
+            "removed in version 3.6. Use `nx.dag.v_structures` or `nx.dag.colliders`\n"
+            "instead.\n"
+        ),
+        category=DeprecationWarning,
+        stacklevel=5,
```
Comment (Schefflera-Arboricola) [inline networkx/algorithms/dag.py:1292]: I'm not sure about this.
(see [comment](https://github.com/networkx/networkx/pull/7398#issuecomment-2161060029))

Comment (dschult) [reply networkx/algorithms/dag.py:1292]: I think it depends on how many function calls occur on the way to calling k_core.

When I run nx3.3 in a mamba environment I don't get any message when calling 
`nx.k_core(MG, core_number=cn)`.
But if I define a function: `def f():return nx.k_core(MG, core_number=cn)` and run `f()`, I do see it. If I bury it deeper (`def g():return f()`, etc for h(), hh(), ...) I still see the warning, but the line shown as causing the warning changes. We have decorators so sometimes it is hard to count how many levels to go up. But I don't understand why running colab would change the number of levels.

Comment (dschult) [reply networkx/algorithms/dag.py:1292]: It seems like `stacklevel=2` works for the DeprecationWarning. That shows the line of the function where `k_core` is called as the location of the warning. A stacklevel of 1 shows the line where the warning is created. I believe a stack level of 3 show the line of a function calling a function that calls `k_core`.
Anyway, that should be good enough and we can make an issue to figure out more about this.   :)

Comment (Schefflera-Arboricola) [reply networkx/algorithms/dag.py:1292]: I think maybe we can use the `skip_file_prefixes` kwarg here. But it was added in v3.12 so if we add it now then the tests running with Python 3.11 and 3.10 will fail. But, we can consider using this in the future.

from the [docs](https://docs.python.org/3/library/warnings.html):

```The skip_file_prefixes keyword argument can be used to indicate which stack
frames are ignored when counting stack levels. This can be useful when you want
the warning to always appear at call sites outside of a package when a constant
stacklevel does not fit all call paths or is otherwise challenging to maintain.
.....
When prefixes are supplied, stacklevel is implicitly overridden to be max(2, stacklevel)
```

Also, I created a [question](https://stackoverflow.com/questions/78620448/inconsistent-behavior-of-warnings-warn) on Stackoverflow about this.

Comment (Schefflera-Arboricola) [reply networkx/algorithms/dag.py:1292]: One suggestion from the [question](https://stackoverflow.com/questions/78620448/inconsistent-behavior-of-warnings-warn) was to use the `warnings.simplefilter("always")` filter which will cause all warnings to always be triggered. But, I don't think we should add it here as it might override other filters that a user might have used. 

Also, adding https://github.com/ipython/ipython/issues/11207 for reference.

Comment (Schefflera-Arboricola) [reply networkx/algorithms/dag.py:1292]: I tried re-running `kcores` in colab and I did see the warning the first time I ran the code(with `stacklevel=5`) but not the next time and next to next time. I think I missed it previously. I think we can merge this PR with `stacklevel=5` @dschult. Thanks!

Comment (dschult) [reply networkx/algorithms/dag.py:1292]: Why stacklevel=5 instead of stacklevel=2 as the SOF question/answer suggested?

Comment (dschult) [reply networkx/algorithms/dag.py:1292]: Also, I think the suggestion of using `warnings.simplefilter("always")` is a suggestion for debugging the correct stacklevel more than it is a suggestion for published code. That is, use that to figure out how it works, but then let the user set their own filter.

Comment (dschult) [reply networkx/algorithms/dag.py:1292]: Looking at this again:  It is true that `stacklevel=5` makes this work while stacklevel=2 or 3 or 4 does not. To figure that out I tried adding code like this in place of the deprecationwarning.
```python
for s in range(8):
    warnings.warn(f"stacklevel={s}", category=DeprecationWarning, stacklevel=s)
```
It shows which stacklevel the warning is displayed. Then when i turn on `warnings.simplefilter("always")` it shows the line of code from each stacklevel. It is pretty clear that the dispatchable decorator adds 3 levels. And IPython adds more layers on top of that when calling each cell.

In any case, `stacklevel=5` is the value we want.

**Review by dschult — APPROVED** [review]
This PR is ready to merge as far as I can tell.
Thanks @Schefflera-Arboricola for all the discussion comments. :)

I approve this PR.

File: networkx/algorithms/dag.py (hunk 11 above)
Comment (Schefflera-Arboricola) [reply networkx/algorithms/dag.py:1292]: yes, I was also only getting the deprecation warning displayed with `stacklevel=5` but I didn't know why. I knew it probably had something to do with decorators (`@nx._dispatchable`, `not_implemented_for`) but I wasn't clear about the details of how and where the `stacklevel` was actually getting increased.

**Review by rossbar — APPROVED** [review]
Excellent work @Schefflera-Arboricola , thanks for scoping out and solving this one! I took the liberty of pushing up a few minor wording/rst formatting tweaks (and rebased on `main` to pull in your fix for the docs CI) but LGTM!
